### PR TITLE
LPD-23250 | LPD-23251 | LPD-23252 | LPD-31826

### PR DESCRIPTION
**Pull request by cdbm**
Pr related to story [LPD-6829](https://liferay.atlassian.net/browse/LPD-6829) Reset scim client provisioning data. Sending this pr now without tests to start review while I am writing the tests. See ticket description for ACs

### CODE REVIEW

**Comment by cdbm** [comment]
ci:test:sf

**Comment by cdbm** [comment]
ci:test:sf

**Comment by cdbm** [comment]
ci:test:relevant

**Comment by ChrisKian** [comment]
Test failures look unrelated.  They seemt o be due to unrelated poshi macros failing, like `HeadlessPortalInstanceAPI.macro` and `ApplicationsMenuHelper.macro`.

File: modules/dxp/apps/scim/scim-configuration-web/src/main/java/com/liferay/scim/configuration/web/internal/portlet/action/SaveScimConfigurationMVCActionCommand.java [hunk 1]
```diff
@@ -155,6 +169,48 @@ else if (Objects.equals(cmd, "reset")) {
...
+
+			Indexer<UserGroup> userGroupIndexer =
+				IndexerRegistryUtil.nullSafeGetIndexer(UserGroup.class);
+
+			ActionableDynamicQuery query =
+				_expandoValueLocalService.getActionableDynamicQuery();
+
+			query.setAddCriteriaMethod(
+				dynamicQuery -> {
+					Property dataProperty = PropertyFactoryUtil.forName("data");
```
Comment (ChrisKian) [inline modules/dxp/apps/scim/scim-configuration-web/src/main/java/com/liferay/scim/configuration/web/internal/portlet/action/SaveScimConfigurationMVCActionCommand.java:184]: Shouldn't this be "data_"?  See https://github.com/liferay/liferay-portal/blob/master/portal-impl/src/com/liferay/portlet/expando/model/impl/ExpandoValueModelImpl.java#L81

File: modules/dxp/apps/scim/scim-configuration-web/src/main/java/com/liferay/scim/configuration/web/internal/portlet/action/SaveScimConfigurationMVCActionCommand.java [hunk 2]
```diff
@@ -155,6 +169,48 @@ else if (Objects.equals(cmd, "reset")) {
...
+
+			query.setPerformActionMethod(
+				(ExpandoValue expandoValue) -> {
+					_expandoRowLocalService.deleteRow(
+						expandoValue.getTableId(), expandoValue.getClassPK());
+
+					ClassName className = _classNameLocalService.getClassName(
+						expandoValue.getClassNameId());
+
+					String classNameString = className.getClassName();
```
Comment (ChrisKian) [inline modules/dxp/apps/scim/scim-configuration-web/src/main/java/com/liferay/scim/configuration/web/internal/portlet/action/SaveScimConfigurationMVCActionCommand.java:197]: We can use Portal.getClassName() instead, see https://github.com/liferay/liferay-portal/blob/master/portal-impl/src/com/liferay/portal/util/PortalImpl.java#L1722-L1734

File: modules/dxp/apps/scim/scim-configuration-web/src/main/java/com/liferay/scim/configuration/web/internal/portlet/action/SaveScimConfigurationMVCActionCommand.java [hunk 3]
```diff
@@ -155,6 +169,48 @@ else if (Objects.equals(cmd, "reset")) {
...
+
+					if (classNameString.equals(User.class.getName())) {
+						userIndexer.reindex(
+							_userLocalService.getUser(
+								expandoValue.getClassPK()));
+					}
+					else if (classNameString.equals(
```
Comment (ChrisKian) [inline modules/dxp/apps/scim/scim-configuration-web/src/main/java/com/liferay/scim/configuration/web/internal/portlet/action/SaveScimConfigurationMVCActionCommand.java:204]: Is there every a situation in SCIM where the `classNameString` won't equal `User.class.getName()` or `UserGroup.class.getName()`?  If those are the only two possible options, then the `else if()` should just be `else`.

File: modules/dxp/apps/scim/scim-configuration-web/src/main/java/com/liferay/scim/configuration/web/internal/portlet/action/SaveScimConfigurationMVCActionCommand.java [hunk 4]
```diff
@@ -155,6 +169,48 @@ else if (Objects.equals(cmd, "reset")) {
 
 			_oAuth2ApplicationLocalService.deleteOAuth2Application(
 				oAuth2Application);
+
+			Indexer<User> userIndexer = IndexerRegistryUtil.nullSafeGetIndexer(
+				User.class);
+
+			Indexer<UserGroup> userGroupIndexer =
+				IndexerRegistryUtil.nullSafeGetIndexer(UserGroup.class);
```
Comment (ChrisKian) [inline modules/dxp/apps/scim/scim-configuration-web/src/main/java/com/liferay/scim/configuration/web/internal/portlet/action/SaveScimConfigurationMVCActionCommand.java:177]: We should only create these `Indexer<>` objects as needed.  With the current logic, you will create two but only use one at most.

File: modules/dxp/apps/scim/scim-configuration-web/src/main/resources/META-INF/resources/portal_settings/scim_configuration.jsp [hunk 5]
```diff
@@ -167,7 +167,7 @@ String oAuth2ApplicationName = GetterUtil.getString(request.getAttribute(ScimWeb
 	}
 
 	var resetSCIMClientData = document.getElementById(
-			'<portlet:namespace />resetSCIMClientData'
+		'<portlet:namespace />resetSCIMClientData'
```
Comment (ChrisKian) [inline modules/dxp/apps/scim/scim-configuration-web/src/main/resources/META-INF/resources/portal_settings/scim_configuration.jsp:170]: This commit should be squashed into the commit which introduced the SF issue.

**Comment by ChrisKian** [comment]
Hey @cdbm, I added a few comments you can go through.  I will handle most the SF related comments myself, but there are some questions I don't know the answers to without testing changes.  Here are the current questions I'd like answered:
1. Looking at [this comment](https://github.com/liferay-appsec/liferay-portal/pull/1874#discussion_r1700566040), is it possible to have a SCIM client configured for both a `User` and `UserGroup` at the same time?  Basically, I'd like to know if we **_need_** to create both `indexer` objects, or if we can just create one to save resources.
2. > Is there ever a situation in SCIM where the `classNameString` won't equal `User.class.getName()` or `UserGroup.class.getName()`?...
3. > Shouldn't this be "`data_`"? See https://github.com/liferay/liferay-portal/blob/master/portal-impl/src/com/liferay/portlet/expando/model/impl/ExpandoValueModelImpl.java#L81

File: modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language.properties [hunk 6]
```diff
@@ -15580,6 +15580,8 @@ scim-generate-access-token=Generate Access Token
 scim-generate-access-token-help=Generate a new access token while establishing a 10 day grace period for previous access tokens before they are invalidated.
 scim-matcher-field=Matcher Field
 scim-matcher-field-help=The matcher field is used to match and associate existing users that are not already associated with a SCIM client. Valid options are schema properties from RFC 7643 that will be matched with appropriate user fields.
+scim-reset-client-data=Reset SCIM Client Provisioning Data
+scim-reset-client-data-help=All the current SCIM Client related data and the generated OAuth 2 tokens is removed. This is necessary for being able to configure a new SCIM Client.
```
Comment (ChrisKian) [inline modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language.properties:15584]: Small typo, it should be `...generated OAuth 2 tokens ARE removed.`

Just pointing this out FYI; I will take care of this.

File: modules/dxp/apps/scim/scim-configuration-web/src/main/java/com/liferay/scim/configuration/web/internal/portlet/action/SaveScimConfigurationMVCActionCommand.java [hunk 7]
```diff
@@ -145,6 +145,17 @@ else if (Objects.equals(cmd, "revoke")) {
...
+			OAuth2Application oAuth2Application =
+				_oAuth2ApplicationLocalService.getOAuth2Application(
+					themeDisplay.getCompanyId(), scimClientId);
+
+			_oAuth2AuthorizationService.revokeAllOAuth2Authorizations(
+				oAuth2Application.getOAuth2ApplicationId());
+
+			_oAuth2ApplicationLocalService.deleteOAuth2Application(
+				oAuth2Application);
+		}
```
Comment (ChrisKian) [inline modules/dxp/apps/scim/scim-configuration-web/src/main/java/com/liferay/scim/configuration/web/internal/portlet/action/SaveScimConfigurationMVCActionCommand.java]: The command `reset` comes before `revoke` alphabetically, so these conditions need reordering.  I will sort this myself, just pointing it out FYI.

File: modules/dxp/apps/scim/scim-configuration-web/src/main/java/com/liferay/scim/configuration/web/internal/portlet/action/SaveScimConfigurationMVCActionCommand.java (hunk 3 above)
Comment (cdbm) [reply modules/dxp/apps/scim/scim-configuration-web/src/main/java/com/liferay/scim/configuration/web/internal/portlet/action/SaveScimConfigurationMVCActionCommand.java:204]: Only user and userGroup are used as for now, i just left the else if to be precise/clear. but i can change it if needed

File: modules/dxp/apps/scim/scim-configuration-web/src/main/java/com/liferay/scim/configuration/web/internal/portlet/action/SaveScimConfigurationMVCActionCommand.java (hunk 4 above)
Comment (cdbm) [reply modules/dxp/apps/scim/scim-configuration-web/src/main/java/com/liferay/scim/configuration/web/internal/portlet/action/SaveScimConfigurationMVCActionCommand.java:177]: we can have both users and user groups related to a scim client. The configurations process are separated from each other as well as the APIs

File: modules/dxp/apps/scim/scim-configuration-web/src/main/java/com/liferay/scim/configuration/web/internal/portlet/action/SaveScimConfigurationMVCActionCommand.java (hunk 3 above)
Comment (ChrisKian) [reply modules/dxp/apps/scim/scim-configuration-web/src/main/java/com/liferay/scim/configuration/web/internal/portlet/action/SaveScimConfigurationMVCActionCommand.java]: That does make sense, however it also adds an unnecessary String comparison, and the next line tells the user exactly what is happening:

> userGroupIndexer.reindex(
> 	_userGroupLocalService.getUserGroup(
> 		expandoValue.getClassPK()));

I'll adjust this in my commits

File: modules/dxp/apps/scim/scim-configuration-web/src/main/java/com/liferay/scim/configuration/web/internal/portlet/action/SaveScimConfigurationMVCActionCommand.java (hunk 1 above)
Comment (ChrisKian) [reply modules/dxp/apps/scim/scim-configuration-web/src/main/java/com/liferay/scim/configuration/web/internal/portlet/action/SaveScimConfigurationMVCActionCommand.java:184]: As you mentioned, "data" works, but "data_" doesn't, even though it is the true column name.

I see other examples where the column ends with an underscore, but it is not included in the `dataProperty`, so your approach is correct.  (See "type_" for the `group_` table)

**Pull request closed without merging**